include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,367 @@ enum E_DESCRIPTOR_TYPE : uint8_t
-	binding.binding = M;
-	binding.stageFlags = static_cast<asset::ICPUSpecializedShader::E_SHADER_STAGE>(asset::ICPUSpecializedShader::ESS_VERTEX | asset::ICPUSpecializedShader::ESS_FRAGMENT);
-	binding.type = X; // It might be an asset::EDT_UNIFORM_BUFFER for instance
-	auto descriptorSetLayout = core::make_smart_refctd_ptr<asset::ICPUDescriptorSetLayout>(&binding, &binding + 1);
-
-	// Assuming that set N = 1, you execute std::move() one second constructor's field of available descriptor set layouts
-	auto pipelineLayout = core::make_smart_refctd_ptr<asset::ICPUPipelineLayout>(nullptr, nullptr, nullptr, std::move(descriptorSetLayout), nullptr, nullptr);
-	\code
-
 	@see IReferenceCounted
 */
 
 template<typename SamplerType>
 class NBL_API IDescriptorSetLayout : public virtual core::IReferenceCounted
 {
-	struct SBindingRedirect
+public:
+	using sampler_type = SamplerType;
+
+	struct SBinding
 	{
+		enum class E_CREATE_FLAGS : uint8_t
+		{
+			ECF_NONE = 0,
+			ECF_UPDATE_AFTER_BIND_BIT = 1u << 1,
+			ECF_UPDATE_UNUSED_WHILE_PENDING_BIT = 1u << 2,
+			ECF_PARTIALLY_BOUND_BIT = 1u << 3
+		};
+
+		uint32_t binding;
+		IDescriptor::E_TYPE type;
+		core::bitflag<E_CREATE_FLAGS> createFlags;
+		core::bitflag<IShader::E_SHADER_STAGE> stageFlags;
+		uint32_t count;
+		// Use this if you want an immutable sampler that is baked into the DS layout itself.
+		// If its `nullptr` then the sampler used is mutable and can be specified while writing the image descriptor to a binding while updating the DS.
+		const core::smart_refctd_ptr<sampler_type>* samplers;
+	};
+
+	// Maps a binding to a local (to descriptor set layout) offset.
+	class CBindingRedirect
+	{
+	public:
 		static constexpr inline uint32_t Invalid = ~0u;
 
-		SBindingRedirect() : bindings(nullptr), offsets(nullptr), count(0ull) {}
+		struct binding_number_t
+		{
+			inline binding_number_t(const uint32_t d) : data(d) {}
+			uint32_t data;
+		};
+
+		struct storage_offset_t
+		{
+			inline storage_offset_t(const uint32_t d) : data(d) {}
+			uint32_t data;
+		};
+
+		inline uint32_t getBindingCount() const { return m_count; }
+
+		// Returns index into the binding property arrays below (including `m_storageOffsets`), for the given binding number `binding`.
+		// Assumes `m_bindingNumbers` is sorted and that there are no duplicate values in it.
+		inline uint32_t searchForBinding(const binding_number_t binding) const
+		{
+			if (!m_bindingNumbers)
+				return Invalid;
+
+			assert(m_storageOffsets && (m_count != 0u));
+
+			auto found = std::lower_bound(m_bindingNumbers, m_bindingNumbers + m_count, binding, [](binding_number_t a, binding_number_t b) -> bool {return a.data < b.data; });
+
+			if ((found >= m_bindingNumbers + m_count) || (found->data != binding.data))
+				return Invalid;
+
+			const uint32_t foundIndex = found - m_bindingNumbers;
+			assert(foundIndex < m_count);
+			return foundIndex;
+		}
+
+		inline binding_number_t getBindingNumber(const uint32_t index) const
```

Comment:
there two probably need better names, something like `findBindingStorageIndex` and `getBindingForStorageIndex`

Reply:
also you should be using `storage_offset_t` instead of the `uint32_t` in both

Reply:
hmm ok I see `storage_offset_t` is for offsets to store individual descriptors.

Reply:
Lets introduce `storage_range_index_t` as a type for storing the index into the storage offset array

Reply:
Resolved.